# Wine

In [3]:
#module import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

#loading data and check columns
wines = load_wine()
print(wines.keys())
print(wines['data'])
print(wines['feature_names'])
print(wines['target_names'])

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])
[[1.423e+01 1.710e+00 2.430e+00 ... 1.040e+00 3.920e+00 1.065e+03]
 [1.320e+01 1.780e+00 2.140e+00 ... 1.050e+00 3.400e+00 1.050e+03]
 [1.316e+01 2.360e+00 2.670e+00 ... 1.030e+00 3.170e+00 1.185e+03]
 ...
 [1.327e+01 4.280e+00 2.260e+00 ... 5.900e-01 1.560e+00 8.350e+02]
 [1.317e+01 2.590e+00 2.370e+00 ... 6.000e-01 1.620e+00 8.400e+02]
 [1.413e+01 4.100e+00 2.740e+00 ... 6.100e-01 1.600e+00 5.600e+02]]
['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']
['class_0' 'class_1' 'class_2']


In [17]:
#sepearting fearture with target
wine_data = wines['data']
wine_target = wines['target']
print(wine_data)
print(wine_target)

[[1.423e+01 1.710e+00 2.430e+00 ... 1.040e+00 3.920e+00 1.065e+03]
 [1.320e+01 1.780e+00 2.140e+00 ... 1.050e+00 3.400e+00 1.050e+03]
 [1.316e+01 2.360e+00 2.670e+00 ... 1.030e+00 3.170e+00 1.185e+03]
 ...
 [1.327e+01 4.280e+00 2.260e+00 ... 5.900e-01 1.560e+00 8.350e+02]
 [1.317e+01 2.590e+00 2.370e+00 ... 6.000e-01 1.620e+00 8.400e+02]
 [1.413e+01 4.100e+00 2.740e+00 ... 6.100e-01 1.600e+00 5.600e+02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]


In [18]:
#spliting train data and test data
train_feature, test_feature, train_target, test_target = train_test_split(wine_data, 
                                                                          wine_target, 
                                                                          test_size=0.3, 
                                                                          random_state=5)

In [20]:
#testing for Decision-Tree Learning
dct = DecisionTreeClassifier(random_state=100)
dct.fit(train_feature, train_target)
dct_predict = dct.predict(test_feature)
print(classification_report(test_target, dct_predict))

              precision    recall  f1-score   support

           0       0.95      0.91      0.93        23
           1       0.89      0.89      0.89        18
           2       0.93      1.00      0.96        13

    accuracy                           0.93        54
   macro avg       0.92      0.93      0.93        54
weighted avg       0.93      0.93      0.93        54



In [21]:
#testing for Random Forest
rdf = RandomForestClassifier(random_state=100)
rdf.fit(train_feature, train_target)
rdf_predict = rdf.predict(test_feature)
print(classification_report(test_target, rdf_predict))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        23
           1       1.00      0.94      0.97        18
           2       0.93      1.00      0.96        13

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



In [22]:
#testing for Support Vector Machine
sptvm = svm.SVC(random_state=100)
sptvm.fit(train_feature, train_target)
sptvm_predict = sptvm.predict(test_feature)
print(classification_report(test_target, sptvm_predict))

              precision    recall  f1-score   support

           0       0.89      0.74      0.81        23
           1       0.60      0.83      0.70        18
           2       0.40      0.31      0.35        13

    accuracy                           0.67        54
   macro avg       0.63      0.63      0.62        54
weighted avg       0.68      0.67      0.66        54



In [9]:
#testing for Stochastic Gradient Decent Classifier
sgdc = SGDClassifier(random_state=100)
sgdc.fit(train_feature, train_target)
sgdc_predict = sgdc.predict(test_feature)
print(classification_report(test_target, sgdc_predict))

              precision    recall  f1-score   support

           0       1.00      0.43      0.60        44
           1       0.71      0.54      0.62        46
           2       0.35      0.71      0.47        35

    accuracy                           0.55       125
   macro avg       0.69      0.56      0.56       125
weighted avg       0.71      0.55      0.57       125



In [23]:
#testing for Logistic Regresssion
lgr = LogisticRegression(random_state=100, max_iter=3000)
lgr.fit(train_feature, train_target)
lgr_predict = sgdc.predict(test_feature)
print(classification_report(test_target, lgr_predict))

              precision    recall  f1-score   support

           0       1.00      0.39      0.56        23
           1       0.82      0.50      0.62        18
           2       0.32      0.85      0.47        13

    accuracy                           0.54        54
   macro avg       0.71      0.58      0.55        54
weighted avg       0.78      0.54      0.56        54



# Conclusion
scikit-learn의 와인데이터에 관해 5가지의 모델(Decision-Tree, Random Forest, Support Vector Machine, SGDC, Logistic Regression)으로 학습시켜 보았다.
와인을 3가지의 class로 분류함에 있어 feature factor에 관해 얼마나 정확한 class를 찾아 낼 수 있는지가 가장 중요함으로 정밀도(precison)을 고려할 수 있다. 위 5가지의 모델링 결과 precison이 가장 높은 모델은 Random Forest이며, 0.98의 정밀도를 보였다.